libscript/src/script-bytecode.hpp

```diff
@@ -0,0 +1,904 @@
+	
+	MC_SCRIPT_DEFINE_BYTECODE(Invoke, "invoke", "hrr*")
+	
+	static void Validate(MCScriptValidateContext& ctxt)
+	{
+		ctxt.CheckMinimumArity(2);
+		
+		ctxt.CheckHandler(ctxt.GetArgument(0));
+		
+		for(uindex_t i = 1; i < ctxt.GetArity(); i++)
+			ctxt.CheckRegister(ctxt.GetArgument(i));
+	}
+	
+	static void Execute(MCScriptExecuteContext& ctxt)
+	{
+		// Resolve the import definition chain for the specified definition.
+		MCScriptInstanceRef t_instance = nil;
+		MCScriptDefinition *t_definition = nil;
+		ctxt.ResolveDefinition(ctxt.GetArgument(0),
+							   t_instance,
+							   t_definition);
+		
+		// The result register is the second argument.
+		uindex_t t_result_reg;
+		t_result_reg = ctxt.GetArgument(1);
+		
+		// The argument list starts at the third argument.
+		uindex_t *t_argument_regs;
+		uindex_t t_argument_count;
+		t_argument_regs = ctxt.GetArgumentList() + 2;
+		t_argument_count = ctxt.GetArgumentCount() - 2;
+		
+		
+		// If the definition kind is a group, then we must select which handler
+		// in the group is appropriate based on the in-mode argument types.
+		if (t_definition -> kind == kMCScriptDefinitionKindDefinitionGroup)
+		{
+			SelectDefinitionFromGroup(ctxt,
+									  t_instance,
+									  static_cast<MCScriptDefinitionGroupDefinition *>(t_definition),
+									  t_argument_regs,
+									  t_argument_count,
+									  t_instance,
+									  t_definition);
+		}
+		
+		// If we have no definition (due to failure), we return,
+		if (t_definition == nil)
+			return;
+		
+		switch(t_definition->kind)
+		{
+			case kMCScriptDefinitionKindHandler:
+			{
+				ctxt.PushFrame(t_instance,
+							   static_cast<MCScriptHandlerDefinition *>(t_definition),
+							   t_result_reg,
+							   t_argument_regs,
+							   t_argument_count);
+			}
+			break;
+				
+			case kMCScriptDefinitionKindForeignHandler:
+			{
+				ctxt.InvokeForeign(t_instance,
+								   static_cast<MCScriptForeignHandlerDefinition *>(t_definition),
+								   t_result_reg,
+								   t_argument_regs,
+								   t_argument_count);
+			}
+			break;
+				
+			default:
+			{
+				__MCScriptUnreachable__("invalid definition kind in invoke");
+			}
+			break;
+		}
+	}
+	
+private:
+	static void SelectDefinitionFromGroup(MCScriptExecuteContext& ctxt,
+										  MCScriptInstanceRef p_instance,
+										  MCScriptDefinitionGroupDefinition *p_group,
+										  uindex_t *p_arguments,
+										  uindex_t p_argument_count,
+										  MCScriptInstanceRef& r_selected_instance,
+										  MCScriptDefinition*& r_selected_definition)
+	{
+		// We use a simple scoring mechanism to determine which method to use. If
+		// input type for an argument is equal to expected type, then this is a
+		// score of zero. If input type for an argument conforms to expected type, then
+		// this is a score of 1. The method with the lowest score is chosen, with
+		// priority given to the first in the list in the case of a tie.
```

Comment:
I still feel like a score tie should be an error, because it makes runtime behaviour sensitive to the order in which the syntax modules happened to be compiled.